Publicize: Adding to Gutenberg Extensibility for Publicize Support

c-shultz opened the pull request:
Fixes #3264 

## Description

To support the current Publicize integration strategy per Jetpack PR Automattic/jetpack#9144, I'm proposing Gutenberg provide extensibility in a couple areas:

- ~~Provide a hook that is triggered _just_ before publish that can trigger plugin callbacks~~
- **Provide the ability to differentiate server requests coming from Gutenberg**
- Provide extendable pre and publish panels (see strategy in #3264)

## Progress so far

~~The progress to date in this PR adds a 'prePublish' action hook that is triggered just before a post is published.~~
- Added `X-WP-Source='Gutenberg'` to request header to differentiate post updates coming from Gutenberg.
- Added `PluginPrePublishPanel` and `PluginPostPublishPanel` to for plugins to add content to pre-publish sidebar or post-publish sidebar respectively. Here's a preview of some test content ("Hello world from post publish." in `PluginPostPublishPanel`:

![image](https://user-images.githubusercontent.com/6209518/38144755-1979601c-33fb-11e8-8cb4-d7bf3a68e044.png)



## TODO items:

- ~~Iterate 'prePublish' iteration to allow for asynchronous calls~~
- ~~Add post-publish plugin extendability using registerPlugin~~ **First iteration done**

## Testing
~~'prePublish' action can be tested in browser console, using `wp.hooks.addAction( 'prePublish', 'publicize', function(){alert("Hello World");});` which should trigger alert when post is published.~~

- `X-WP-Source='Gutenberg'` header in place, Gutenberg requests can be picked up on the server side by checking `'Gutenberg' === $_SERVER['HTTP_X_WP_SOURCE']`
-Using the experimental `Plugin<Pre/Post>PublishPanel`s:
```jsx
import { Slot, Fill } from '@wordpress/components';
function PluginPrePublishPanel( { children } ) {
	return (
		<Fill name={ SLOT_NAME } >
			{ children }
		</Fill>
	);
}
PluginPrePublishPanel.Slot = () => (
	<Slot name={ SLOT_NAME } />
);
```



## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code has proper inline documentation.

c-shultz committed:
Gutenberg: Creating prePublish action

Creates a 'prePublish' action hook that is triggered
right before a post is published.

c-shultz commented:
@gziolo , I wanted to get this PR in front of you. I haven't yet added the post-publish functionality that we discussed, but I should get it in this week.

atimmer commented:
For the UI extensibility, I think we should follow the same pattern established for a sidebar:

```jsx
const { Fragment } = wp.element;
const { PrePublish } = wp.editPost.__experimental;
const { registerPlugin } = wp.plugins;

const Component = () => (
	<Fragment>
                <PrePublish name="jetpack">
                     { /* Render Jetpack's pre-publish settings here */ }
                </PrePublish>
	</Fragment>
);

registerPlugin( 'plugin-name', {
	render: Component,
} );
```

For pointers on how to implement this you could take a look at the Sidebar PR: https://github.com/WordPress/gutenberg/pull/5430/. Maybe there should also be a `PostPublish` component.

Related: https://github.com/WordPress/gutenberg/pull/5767.

gziolo commented:
> Provide a hook that is triggered just before publish that can trigger plugin callbacks

Is there anything specific that should be executed on the frontend that can't be performed on the server? In this case you are adding a hook after `REQUEST_POST_UPDATE` action was dispatched. This effect also dispatches a few other actions. All those actions lead to the state change which makes it possible to observe for change using `data` module and [select](https://github.com/WordPress/gutenberg/tree/master/data#select-storename-string--object), [withSelect](https://github.com/WordPress/gutenberg/tree/master/data#withselect-mapselecttoprops-function--function) or [subscribe](https://github.com/WordPress/gutenberg/tree/master/data#subscribe-function).

I think @aduth had some good idea how it all can be handled using only effects and data store.

aduth commented:
>Gutenberg must provide extendablitity in a couple areas:

"Must" is a strong word, implying there is no other alternative. Based on https://github.com/Automattic/jetpack/pull/9144 , I'm assuming this is so that the post can be flagged as not being shared for Publicize. Why can't this occur earlier in the editor lifecycle? Why can't this be the default to not publicize by default? Why can't Jetpack hook into server API save filters? It seems to promote a poor user experience to _block_ editor saving so that a plugin's own network request can go through. And if it's not blocking, which is the case as implemented here, there's a very likely risk for race conditions between the two requests.

c-shultz commented:
>"Must" is a strong word, implying there is no other alternative. Based on Automattic/jetpack#9144 , I'm assuming this is so that the post can be flagged as not being shared for Publicize. Why can't this occur earlier in the editor lifecycle? Why can't this be the default to not publicize by default?

Hey, @aduth. Very valid points. The behavior I'm looking at implementing here is definitely not a hard requirement overall, just a piece of what's needed to support the current strategy in Automattic/jetpack#9144. 

We had some discussion on the original Jetpack issue about this approach: Automattic/jetpack#9039. I'm not particularly happy with the current solution of generating the request right before publish. As far as I know, there's not currently any reliable way to differentiate a post being published from Gutenberg vs. a post being published from other sources (classic editor, mobile, API, etc.).

Having this action occur earlier in the editor lifecycle may be the best approach if we don't want to hold the "Publish" action while it finishes. The downside of this is it increases the chance that a post is "flagged" and then is later published through other means. If a post is flagged, and then a user publishes elsewhere, then it will not be publicized. A definite corner case, but it'll happen sometimes. I wonder how much we really have to worry about it...

aduth commented:
Yeah, to me, the more concerning thing is that we feel the need to introduce a different behavior for Gutenberg posts vs. other sources.

Another consideration per your point that there's no way to distinguish a Gutenberg point: Maybe a plugin could add such a distinguishing marker? There's a few steps of post preparation that occur [when the editor loads](https://github.com/WordPress/gutenberg/blob/master/lib/client-assets.php).  Maybe one of those could be adding some (temporary?) meta value marking the post as being edited in Gutenberg. I'm not sure this ought to be a core behavior, for the reason mentioned above (distinguishing Gutenberg should at the very least not be a "blessed" behavior).

Even if there were to be some pre-publish hook, I don't think an action is necessarily the best fit. As mentioned previously it's prone to race conditions. I imagine it would be more an array of promises that a plugin could filter into, and is passed through to [`Promise.all`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all) prior to the save. To clarify, I think this is a worst-case (last-resort) implementation.

c-shultz commented:
>Is there anything specific that should be executed on the frontend that can't be performed on the server? 

@gziolo, the current approach in Automattic/jetpack#9144 needs to have a post flagged before it is published. That PR provides a REST endpoint to do just that. The crux of what I need is the ability to differentiate between a post being published from Gutenberg and post being published from other sources.

c-shultz commented:
>Another consideration per your point that there's no way to distinguish a Gutenberg point: Maybe a plugin could add such a distinguishing marker? There's a few steps of post preparation that occur when the editor loads. Maybe one of those could be adding some (temporary?) meta value marking the post as being edited in Gutenberg.

This is definitely doable, but I don't think 100% foolproof. A post could be flagged right when Gutenberg opens up and (if the flag is set up to expire after a set time) it could even be refreshed periodically using the [Heartbeat API](https://developer.wordpress.org/plugins/javascript/heartbeat-api/). Unfortunately that still leaves open the possibility that a user opens the post in Gutenberg, quickly publishes via some other means, and then finds that the post has not been shared.

I don't have a very good perspective on if that's too small of a corner case to worry about.

>to me, the more concerning thing is that we feel the need to introduce a different behavior for Gutenberg posts vs. other sources.

I wonder if there's any chance other plugins will have a similar need. This need is coming about because the design concept takes advantage of Gutenberg's post-publish panel to give the user control over an action that normally happens automatically upon publish. If Jetpack takes advantage of that flow, will there be others to follow, or is this really a once off?...

Speaking from a perspective of only making this one feature work, it would be very nice to get a 'post_publish_available' key in the post request to trigger special back-end actions whenever Gutenberg publishes. I could totally avoid all the kludge of trying to slip another request in before publish occurs. The post-publish view does potentially introduce a new way of splitting up actions that might have happened automatically, so maybe there would be justification for introducing this to allow for plugins to take advantage of the new flow while preserving backwards compatibility?

I definitely don't presume to have any perspective on the overall architectural goals. 😃 I'm just brainstorming a way to make this work in a clean way.

Venutius commented:
There are a number of plugins that check aspects of posts for missing elements and intervene during the publication process to prevent for example, the publication of a post with a duplicate title of missing featured image, Ideally these plugins could provide support for Gutenberg and to my mind you'd need to run different code depending on if Gutenberg is the editor or not. I think we need that level of granularity.

gziolo commented:
> There are a number of plugins that check aspects of posts for missing elements and intervene during the publication process to prevent for example, the publication of a post with a duplicate title or missing featured image, Ideally these plugins could provide support for Gutenberg and to my mind you'd need to run different code depending on if Gutenberg is the editor or not. I think we need that level of granularity.

The same code is used to process posts in both the classic editor and Gutenberg. Whatever works at the moment, it is going to work the same in Gutenberg. The fact that we are using REST API doesn't mean it works differently to whatever is now in the classic editor. If you prevent the post to be updated then you just need to make sure that API response contains the feedback that can be presented to the user in Gutenberg. 

> Speaking from a perspective of only making this one feature work, it would be very nice to get a 'post_publish_available' key in the post request to trigger special back-end actions whenever Gutenberg publishes. I could totally avoid all the kludge of trying to slip another request in before publish occurs. The post-publish view does potentially introduce a new way of splitting up actions that might have happened automatically with publish. Maybe there would be justification for introducing this to allow for plugins to take advantage of the new flow while preserving backward compatibility?

@pento recently added a shim to override `wp.apiRequest`, allowing HTTP/1.0 emulation. It uses `X-HTTP-Method-Override: METHOD` header included in the request. I think the cleanest way to make it easy to distinguish between Gutenberg and everything else is to append another header which will open this possibility. I did a quick inspection of the request headers in Gutenberg and I noticed the following:
> X-Requested-With: XMLHttpRequest
> X-WP-Nonce: 01249349b5

We might add something along those lines:
> X-WP-Source: Gutenberg

Having that exposed, you can use PHP hooks to add as many changes as required for the plugin to make such flow work.

/cc @adamsilverstein

c-shultz committed:
Revert "Gutenberg: Creating prePublish action"

This reverts commit 81696afec2bd3c37e78d944b73c7978b06431ad2.
Using action hooks and an extra request was deemed too much kludge
just to differentiate publishing from Gutenberg. Another iteration
is incoming...

c-shultz committed:
Extensibility: X-WP-Source='Gutenberg' header added

As a method for differentiating requests from Gutenberg editor,
a header keyed 'X-WP-Source' is being added and set to 'Gutenberg'.

c-shultz commented:
@gziolo, That's a great idea. I just reverted the last commit and used your suggestion instead: `X-WP-Source='Gutenberg'`

gziolo commented:
Glad to hear it simplifies things. I would double check if that shim is loaded when Gutenberg plugin is enabled vs when the post is edited with Gutenberg. It might be the first option as far as I ubderstood when talking to @pento.

c-shultz committed:
Extensiblity: Adding pre and post publish slots

Adding slots for pre-publish sidebar and post-publish sidebar to be used
with `registerPlugin` for plugins to add content to the bottom of these
sidebars. The current plan is to use the post-publish sidebar will be used
for Jetpack's Publicize feature.

gziolo commented on edit-post/README.md:
````diff
@@ -96,3 +98,28 @@ An SVG React element to be rendered to the left of the menu item label.
+#### Usage
+
+```jsx
+<PluginPrePublishPanel>
````
Should be Post not Pre 😃

c-shultz replied:
Fixed 😃

gziolo commented on edit-post/components/plugin-post-publish-panel/index.js:
```diff
@@ -0,0 +1,52 @@
+ * @see /edit-post/README.md 
+ *
+ * @file   This files defines the PluginPostPublishPanel extension
+ * @author ChrisShultz
```
We don't use `@author` in the project.

c-shultz replied:
Good to know. I'll kill it. 

Should I be consulting any standards other than what's here? - https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript

c-shultz replied:
Fixed

gziolo commented on edit-post/components/plugin-post-publish-panel/index.js:
```diff
@@ -0,0 +1,52 @@
+ */
+import { Slot, Fill } from '@wordpress/components';
+
+/**
```
No need it to put it here if there are no deps.

c-shultz replied:
Thanks. I removed the unnecessary `@see /edit-posts/README.md`

gziolo commented on edit-post/components/plugin-post-publish-panel/index.js:
```diff
@@ -0,0 +1,52 @@
+ *
+ * @return {WPElement} Plugin sidebar fill.
+ */
+function PluginPostPublishPanel( { children } ) {
```
Should we introduce factory method to remove code duplication (similar to the new React Context AP)?:
```js
const createSlotFill = ( slotName ) => {
	function Component( { children } ) {
		return (
			<Fill name={ slotName } >
				{ children }
			</Fill>
		);
	}
	
	Component.displayName = slotName;

	Component.Slot = () => (
		<Slot name={ slotName } />
	);

	return Component;
};
```
then in this file:
```js
export default createSlotFill( 'PluginPostPublishPanel' );
```
/cc @aduth

c-shultz replied:
That's a good idea. No sense duplicating code. I can only assume there will be the need for more slots as Gutenberg evolves.

I wonder where `createSlotFill` should site. Potentially move it into a subfolder in the top level 'components' directory?

c-shultz replied:
Awesome! I'll look at incorporating this.

c-shultz committed:
Extensibility: Whitespace cleanup

Cleaning up some leading and trailing whitespace issues.

gziolo commented on editor/components/post-publish-panel/prepublish.js:
```diff
@@ -11,6 +11,7 @@ import PostVisibility from '../post-visibility';
 import PostVisibilityLabel from '../post-visibility/label';
 import PostSchedule from '../post-schedule';
 import PostScheduleLabel from '../post-schedule/label';
+import PluginPrePublishPanel from '../../../edit-post/components/plugin-pre-publish-panel';
```
`editor` module shouldn't depend on `edit-post`.
@youknowriad, should we move the PublishPanel to `edit-post`?

c-shultz replied:
What's the high-level difference between which content is in `edit-post` vs `editor`?

youknowriad replied:
`editor` is a generic module for building any editor layout
`edit-post` is the layout built for editing posts/pages/cpts
we can imagine `edit-p2` `edit-template`...

A simple solution here would be to use the  `children` prop.

c-shultz replied:
>editor is a generic module for building any editor layout
edit-post is the layout built for editing posts/pages/cpts
we can imagine edit-p2 edit-template...

Aha, I see. It makes sense to me now why we can't have `editor` depending on `edit-post`.

>A simple solution here would be to use the children prop.

Are you suggesting including `PluginPrePublishPanel` in the `edit-post` layout (/edit-post/component/layout/index.js). Maybe something like the following? --
```jsx
			{ publishSidebarOpened && (
				<PostPublishPanel
					onClose={ closePublishSidebar }
					forceIsDirty={ hasActiveMetaboxes }
					forceIsSaving={ isSaving }
				>
					<PluginPrePublishPanel.Slot/>
				</PostPublishPanel>
			) }
```

And then of course, passing it on the way down so it gets included in the right place in the `editor` layout?

That would keep `editor` as generic as possible. However, I could see the advantage of including the `PluginPrePublishPanel` Slot in `editor`. I'm guessing this extensibility would be useful for any `editor` implementation.

youknowriad replied:
> That being said, I could see the advantage of including the PluginPrePublishPanel Slot in editor. I'm guessing this extensibility would be useful for any editor implementation.

I don't think we should force an extensibility pattern like this to all editors. Saving a template is not the same as saving a post. I'd prefer the former (like the example you shown)

gziolo replied:
Yes, `children` seems like the way to go 👍

c-shultz replied:
@gziolo, I incorporated the new createSlotFill! Thanks!

I also made the change to remove the `edit-post` dependency in `editor`. I couldn't use strictly `children` since the top level `editor` component `<PostPublishPanel>` contains both the 'pre' and the 'post' components. I created `prePublishExtension` and `postPublishExtension` for this purpose instead to route the plugin slots down to right places.

gziolo reviewed (commented):
I left a few early comments.

gziolo commented:
@c-shultz, there went something wrong with merging your changes with the `master` branch. I personally use `git rebase origin/master` to avoid having this kind of issues.

c-shultz commented:
>there went something wrong with merging your changes with the master branch. I personally use git rebase upstream/master to avoid having this kind of issues.

Thanks for pointing that out, @gziolo. I'm going to see if I can get that cleaned up. I need to get my brain out of SVN mode. 😄

adamsilverstein commented:
@gziolo 

> There are a number of plugins that check aspects of posts for missing elements and intervene during the publication process to prevent for example, the publication of a post with a duplicate title or missing featured image

To expand on this, sometimes we want to intervene before the publish button is even clicked, for example, imagine I don't want posts to be publishable until the author has added a featured image. So I want to disable the publish button and add a message above it that says 'add a featured image to publish'. once the featured image is added, the publish button should be enabled and the message should go away. Similarly, published posts would disable 'update' if you tried to remove their featured image. Is this type of extensibility currently possible, can I do this in my plugin or theme?

> We might add something along those lines:

>X-WP-Source: Gutenberg

>Having that exposed, you can use PHP hooks to add as many changes as required for the plugin to make such flow work.

Adding a unique header for Gutenberg requests is a great idea.

aduth commented:
@adamsilverstein Viewed another way, this is not an intervention at publish time, this is an ongoing requirement of the post's state which is satisfied at a known point in time: when the featured image is assigned.

It's not currently possible to achieve to my knowledge, though I would imagine it relating to the behavior of the [`isEditedPostPublishable`](https://github.com/WordPress/gutenberg/blob/d78163dedcdb38cbb98cb94dc16908f35e7d1115/editor/store/selectors.js#L255-L266), either directly or, perhaps more preferably, as an abstraction of post requirements which are handled by the core editor code. The benefit to the abstraction is being able to enforce a consistent and blessed UX for disabling publishing (rather than each plugin rolling their own).

c-shultz committed:
Extensibility: documentation clean up

Fixing typos and clearing out unecessary DocBlock entries.

adamsilverstein commented:
@aduth yea, being able to extend `isEditedPostPublishable` may be the right direction. Plus some sort of slot fill in the publish areas so the reason for the disabled update/publish button is clear. We can track this use case in another issue if its outside the scope here.

c-shultz commented:
@gziolo (or others), can you think of any useful unit testing that could be done here with the current content?

c-shultz commented on edit-post/README.md:
```diff
@@ -7,7 +7,9 @@ Refer to [the plugins module documentation](../plugins/) for more information.
-They can be found in the global variable `wp.editPost` when defining `wp-edit-post` as a script dependency.
+They can be found in the global variable `wp.editPost` when defining `wp-edit-post` as a script dependency. 
+
+**Note:** Plugin scripts should be enqueued to footer by setting `in_footer` option to true ( [see wp_enqueue_script documentation](https://developer.wordpress.org/reference/functions/wp_enqueue_script/) ).
```
@gziolo, Is `in_footer` the best way to handle this? I burned some time chasing down "Target container is not a DOM element" error that appeared whenever the wp-edit-post script was included. Putting it in the footer resolved the issue. 
![not in footer error](https://user-images.githubusercontent.com/6209518/38145246-39b80174-33fd-11e8-959a-044553e636be.PNG)

gziolo replied:
Not sure to be honest. I will investigate it later this week. Ideally, it should be enough to set edit-post as a dependency of your plugin.

c-shultz replied:
The content of my script seemed to be totally irrelevant. My best guess is that edit-post has a dependency that was only there in time if I put it in the footer.

adamsilverstein replied:
we may be injecting some scripts on action hooks without regard to dependencies (ie without wp_enqueue_script)? that could cause this.

gziolo replied:
It has been fixed, you just need to make sure you provide all dependencies when calling `wp_enqueue_script`.

gziolo commented on editor/components/post-publish-panel/prepublish.js:
```diff
@@ -29,6 +30,7 @@ function PostPublishPanelPrepublish() {
 			] }>
 				<PostSchedule />
 			</PanelBody>
+			{ <PluginPrePublishPanel.Slot /> }
```
The same comment about curly braces applies.

gziolo reviewed (commented):
> can you think of any useful unit testing that could be done here with the current content?

Slot & Fill are covered with tests, so it might be not necessary to add new tests with what we have at the moment. It seems to be fine as it is.

c-shultz commented:
@gziolo, I'm being asked to change my direction on the Publicize integration to use the pre-publish instead of post-publish. That won't affect the publish sidebar extensiblity approach here - but it does raise a new need ---

In the classic editor, the Publicize configuration (connection list, and social posting title) is passed in the form field header:
![image](https://user-images.githubusercontent.com/6209518/38338316-f55120f4-381d-11e8-9de5-cddddfa02e7f.png)

For doing this from pre-publish, I need to look into passing the same type of data in the with the POST call to the /wp/v2/posts/<post_id> endpoint. It looks like the `meta` argument is the right place to do this. Is there any pre-existing method for setting post meta from a non-block plugin (like Publicize)? I would love to be able to do something like:

```jsx
setPostMeta( {
   wpas_title: 'custom title',
   wpas_connections: {
      ...
   }
} );
```

c-shultz commented on editor/components/post-publish-panel/postpublish.js:
```diff
@@ -83,6 +84,7 @@ class PostPublishPanelPostpublish extends Component {
 						</ClipboardButton>
 					</div>
 				</PanelBody>
+				{ <PluginPostPublishPanel.Slot /> }
```
✔️

c-shultz committed:
Publish sidebar extensibility: cleaning up obsolete syntax

Removing obsolete curly brackets around PluginPostPublishPanel.Slot and PluginPrePublishPanel.Slot components.

c-shultz committed:
Publish sidebar extensibility: removing edit-post editor dependency

Now passing pre/post publish plugin slot using new prop and children prop
instead of referencing it directly from within editor.

props @youknowriad

c-shultz committed:
Publish sidebar extensibility: utilizing new createSlotFill factory

Replacing repetitious code with newly created createSlotFill factory function
for PluginPrePublishPanel and PluginPostPublishPanel slots.

props @gziolo

gziolo committed:
Update README.md

gziolo committed:
Update compat.php

gziolo committed:
Merge branch 'master' into try/publicize-integration

gziolo commented:
Work continues in #6798.
